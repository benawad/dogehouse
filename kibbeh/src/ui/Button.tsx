import React, {
  ButtonHTMLAttributes,
  DetailedHTMLProps,
  ReactNode,
} from "react";
import { Spinner } from "./Spinner";

const sizeClassnames = {
  big: "py-2 px-6 text-sm rounded-lg",
  small: "px-2 py-1 text-sm rounded-md",
  tiny: "px-1 text-sm rounded-5",
};

const colorClassnames = {
  primary:
    "text-button bg-accent hover:bg-accent-hover disabled:text-accent-disabled disabled:bg-accent-hover",
  secondary:
    "text-button bg-primary-700 hover:bg-primary-600 disabled:text-primary-300",
  "secondary-800":
    "text-button bg-primary-800 hover:bg-primary-600 disabled:text-primary-300",
  "primary-300":
    "text-button bg-primary-700 hover:bg-primary-600 disabled:text-primary-300",
  transparent: "text-button bg-transparent",
  "accent-secondary":
    "text-button bg-secondary hover:bg-secondary-washed-out disabled:text-secondary-washed-out",
};

export type ButtonProps = DetailedHTMLProps<
  ButtonHTMLAttributes<HTMLButtonElement>,
  HTMLButtonElement
> & {
  size?: keyof typeof sizeClassnames;
  color?: keyof typeof colorClassnames;
  loading?: boolean;
  icon?: ReactNode;
  transition?: boolean;
};

export const Button: React.FC<ButtonProps> = ({
  children,
  size = "big",
  color = "primary",
  disabled,
  loading,
  icon,
  className = "",
  transition,
  ...props
}) => {
  return (
    <button
      disabled={disabled || loading}
      className={`flex outline-none focus:ring-4 focus:ring-${color} ${
        sizeClassnames[size]
      } ${transition ? `transition duration-200 ease-in-out` : ``} ${
        colorClassnames[color]
      } font-bold flex items-center justify-center ${className}`}
      data-testid="button"
      {...props}
    >
      <span className={loading ? "opacity-0" : `flex items-center`}>
        {icon ? <span className={`mr-2 items-center`}>{icon}</span> : null}
        {children}
      </span>
      {loading ? (
        <span className={`absolute`}>
          <Spinner size={size === "small" ? "2" : "4"} />
        </span>
      ) : null}
    </button>
  );
};
