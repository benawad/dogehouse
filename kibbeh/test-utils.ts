import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import "mutationobserver-shim";

const customRender = (ui: any, options = {}) =>
  render(ui, {
    ...options,
  });

export * from "@testing-library/react";

export { customRender as render };
