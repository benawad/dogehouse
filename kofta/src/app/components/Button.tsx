import React, { useState } from "react";
import { Spinner } from "./Spinner";
import "../../css/doge-button.css";

export const Button: React.FC<
  React.DetailedHTMLProps<
    React.ButtonHTMLAttributes<HTMLButtonElement>,
    HTMLButtonElement
  > & {
    variant?: "default" | "small" | "slim" | "follow";
    color?: "default" | "red" | "secondary";
    loading?: boolean;
    dogeProbability?: number;
  }
> = ({
  children,
  loading,
  disabled,
  color = "default",
  variant = "default",
  dogeProbability,
  ...props
}) => {
  const [hasDoge, setHasDoge] = useState(false);
  const [dogeXAnchor, setDogeXAnchor] = useState(0);
  const [dogeHideTimeout, setDogeHideTimeout] = useState<NodeJS.Timeout | null>(
    null
  );
  const [useTopDoge, setUseTopDoge] = useState(false);
  const [useGlassesDoge, setUseGlassesDoge] = useState(false);

  const maybeShowDoge = () => {
    if (hasDoge) return;
    if (dogeHideTimeout) {
      clearTimeout(dogeHideTimeout);
      setDogeHideTimeout(null);
      setHasDoge(true);
    } else if ((dogeProbability ?? 0) > Math.random()) {
      setDogeXAnchor(Math.random() * 0.8 + 0.1);
      setHasDoge(true);
      setUseTopDoge(Math.random() < 0.5);
      setUseGlassesDoge(Math.random() < 0.2);
    }
  };

  const hideDoge = () => {
    if (!hasDoge) return;
    setHasDoge(false);
    const timeout = setTimeout(() => {
      setDogeHideTimeout(null);
    }, 200);
    setDogeHideTimeout(timeout);
  };

  const onMouseOver =
    (dogeProbability ?? 0) > 0
      ? (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
          maybeShowDoge();
          if (props.onMouseOver) props.onMouseOver(event);
        }
      : props.onMouseOver;

  const onMouseLeave =
    (dogeProbability ?? 0) > 0
      ? (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
          hideDoge();
          if (props.onMouseLeave) props.onMouseLeave(event);
        }
      : props.onMouseLeave;

  return (
    <button
      {...props}
      onMouseOver={onMouseOver}
      onMouseLeave={onMouseLeave}
      disabled={loading || disabled}
      className={`
          rounded capitalize outline-none w-full flex items-center justify-center text-center text-white button-fix
          ${
            color === "secondary"
              ? "bg-simple-gray-3a hover:bg-simple-gray-45"
              : color === "red"
              ? "bg-red-600 hover:bg-red-400"
              : "bg-blue-500 hover:bg-blue-400"
          }
          ${
            variant === "small"
              ? "py-1 px-2 w-max"
              : variant === "slim"
              ? "max-w-md ml-auto mr-auto py-2.5 px-12"
              : "py-2.5 px-1"
          }
          ${props.className}
        `}
    >
      {loading ? <Spinner /> : null}
      {children}
      {(dogeProbability ?? 0) > 0 ? (
        <>
          <div
            className={`button-doge button-doge-top ${
              hasDoge && useTopDoge ? "button-doge-shown" : ""
            } ${useGlassesDoge ? "button-doge-cooler" : ""}`}
            style={{
              left: `${dogeXAnchor * 100}%`,
            }}
          />
          <div
            className={`button-doge button-doge-bottom ${
              hasDoge && !useTopDoge ? "button-doge-shown" : ""
            } ${useGlassesDoge ? "button-doge-cooler" : ""}`}
            style={{
              left: `${dogeXAnchor * 100}%`,
            }}
          />
        </>
      ) : (
        <></>
      )}
    </button>
  );
};
