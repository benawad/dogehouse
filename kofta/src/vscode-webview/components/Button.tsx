import React from "react";
import { tw } from "twind";
import { Spinner } from "./Spinner";

const colorToBackground = {
  red: "bg-buttonRed hover:bg-buttonHoverRed",
  secondary: "bg-buttonSecondary hover:bg-buttonHoverSecondary",
};

export const Button: React.FC<
  React.DetailedHTMLProps<
    React.ButtonHTMLAttributes<HTMLButtonElement>,
    HTMLButtonElement
  > & {
    variant?: "default" | "small" | "slim" | "follow";
    color?: "default" | "red" | "secondary";
    loading?: boolean;
  }
> = ({
  children,
  style,
  loading,
  disabled,
  color = "default",
  variant = "default",
  ...props
}) => {
  return (
    <button
      disabled={loading || disabled}
      style={{
        borderRadius: 4,
        outline: "1px solid transparent",
        outlineOffset: 2,
        color:
          color === "secondary"
            ? "var(--vscode-button-secondaryForeground)"
            : "var(--vscode-button-foreground)",
        padding: `10px var(--input-padding-horizontal)`,
        fontSize: `var(--vscode-font-size)`,
        textTransform: "capitalize",
        ...(variant === "small" ? { padding: "4px 8px", width: "unset" } : (variant === "slim" ? { marginLeft: "auto", marginRight: "auto", padding: "10px 48px" } : {}) ),
        ...style,
      }}
      className={tw`w-full flex items-center justify-center text-center ${variant === "slim" ? "max-w-md" : ""} ${
        colorToBackground[color as keyof typeof colorToBackground] ||
        "bg-button hover:bg-buttonHover"
      } `}
      {...props}
    >
      {loading ? <Spinner /> : children}
    </button>
  );
};
