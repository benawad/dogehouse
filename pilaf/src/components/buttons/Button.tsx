import React from "react";
import {
  Image,
  ImageSourcePropType,
  StyleSheet,
  Text,
  TouchableOpacity,
  ViewProps,
  ViewStyle,
} from "react-native";
import {
  colors,
  paragraphBold,
  radius,
  smallBold,
} from "../../constants/dogeStyle";
import { Spinner } from "../Spinner";

export type ButtonProps = ViewProps & {
  iconSrc?: ImageSourcePropType;
  size?: "big" | "small";
  color?: "primary" | "secondary";
  loading?: boolean;
  disabled?: boolean;
  title;
  onPress?: () => void;
};

export const Button: React.FC<ButtonProps> = ({
  size = "big",
  color = "primary",
  disabled = false,
  loading = false,
  iconSrc,
  title,
  ...props
}) => {
  let colorStyle = color === "primary" ? primaryStyle : secondaryStyle;
  let big = size === "big";
  let styleSize = big ? bigStyle : smallStyle;

  return (
    <TouchableOpacity
      style={[
        disabled ? colorStyle.disabled : colorStyle.default,
        styleSize,
        props.style,
      ]}
      onPress={props.onPress}
      disabled={disabled || loading}
    >
      {iconSrc && (
        <Image
          source={iconSrc}
          style={[
            { tintColor: colors.text, marginRight: 10, width: 20 },
            loading && { opacity: 0 },
          ]}
        />
      )}
      <Text
        style={[
          big ? colorStyle.text : colorStyle.textSmall,
          loading && { opacity: 0 },
        ]}
      >
        {title}
      </Text>
      {loading && (
        <Spinner style={{ position: "absolute" }} size={big ? "m" : "s"} />
      )}
    </TouchableOpacity>
  );
};

const containerBase: ViewStyle = {
  alignSelf: "baseline",
  flexDirection: "row",
  justifyContent: "center",
  alignItems: "center",
};

const primaryStyle = StyleSheet.create({
  text: {
    ...paragraphBold,
  },
  textSmall: {
    ...smallBold,
  },
  default: {
    ...containerBase,
    backgroundColor: colors.accent,
  },
  disabled: {
    ...containerBase,
    backgroundColor: colors.accentDisabled,
  },
});

const secondaryStyle = StyleSheet.create({
  text: {
    ...paragraphBold,
  },
  textSmall: {
    ...smallBold,
  },
  default: {
    ...containerBase,
    backgroundColor: colors.primary700,
  },
  disabled: {
    ...containerBase,
    backgroundColor: colors.primary300,
  },
});

const smallStyle: ViewStyle = {
  borderRadius: radius.s,
  paddingHorizontal: 10,
};

const bigStyle: ViewStyle = {
  borderRadius: radius.m,
  paddingHorizontal: 38,
  paddingVertical: 8,
  height: 38,
};
