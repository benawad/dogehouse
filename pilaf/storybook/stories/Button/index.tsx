import React from 'react';
import PropTypes from 'prop-types';
import { TouchableHighlight } from 'react-native';

export default function Button({ onPress, children }) {
  return <TouchableHighlight onPress={onPress}>{children}</TouchableHighlight>;
}

Button.defaultProps = {
  children: null,
  onPress: () => {},
};

Button.propTypes = {
  children: PropTypes.node,
  onPress: PropTypes.func,
};
