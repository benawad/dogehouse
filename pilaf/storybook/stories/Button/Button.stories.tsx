import { action } from '@storybook/addon-actions';
import { text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';
import React from 'react';
import { Text } from 'react-native';
import Button from '.';
import CenterView from '../CenterView';

storiesOf('Button', module)
  .addDecorator((getStory) => <CenterView>{getStory()}</CenterView>)
  .add('with text', () => (
    <Button onPress={action('clicked-text')}>
      <Text>{text('Button text', 'Hello Button')}</Text>
    </Button>
  ))
  .add('with some emoji', () => (
    <Button onPress={action('clicked-emoji')}>
      <Text>😀 😎 👍 💯</Text>
    </Button>
  ));
