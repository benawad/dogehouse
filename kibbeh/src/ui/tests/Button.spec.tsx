import React from "react";

import { render } from "../../../test-utils";
import { Button } from "../Button";

describe("Button", () => {
  it("should render a button with CLick me text", () => {
    const { getByText } = render(<Button>Click me</Button>);
    const text = getByText("Click me");

    expect(text).toBeVisible();
  });

  it("should render a disabled button if loading prop is passed", () => {
    const { getByText } = render(<Button loading={true}>Click me</Button>);
    const text = getByText("Click me");

    expect(text.closest("button")).toBeDisabled();
  });

  it("should render a disabled button if disabled prop is passed", () => {
    const { getByTestId } = render(<Button disabled={true}>Click me</Button>);
    const button = getByTestId("button");

    expect(button).toBeDisabled();
  });

  it("should render an icon if icon props is passed", () => {
    const { getByText, container } = render(
      <Button icon={<div className="flex">Fake icon</div>}>Click me</Button>
    );
    const icon = getByText("Fake icon");

    expect(icon).toBeVisible();
  });

  it("should match snapshot", () => {
    const { getByTestId } = render(<Button disabled={true}>Click me</Button>);
    const button = getByTestId("button");

    expect(button).toMatchSnapshot();
  });
});
