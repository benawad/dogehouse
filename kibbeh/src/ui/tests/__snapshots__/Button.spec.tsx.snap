// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match snapshot 1`] = `
<button
  class="flex outline-none focus:ring-4 focus:ring-primary py-2 px-6 text-sm rounded-lg  text-button bg-accent hover:bg-accent-hover disabled:text-accent-disabled disabled:bg-accent-hover font-bold flex items-center justify-center "
  data-testid="button"
  disabled=""
>
  <span
    class="flex items-center"
  >
    Click me
  </span>
</button>
`;
